### DISCUSSION TITLE
Computing satellite orbits with multiple TLEs

### DISCUSSION THREAD

**Discussion by syk-hugo** [skyfielders/python-skyfield#746] [Q&A] [answered]
Hello, 
I'm trying to instantiate one satellite from various TLEs as its orbit varies and needs multiple TLEs to be fully described. 
Would it be possible to do that with the "EarthSatellite" function from skyfield ? 
Thank you in advance for your help.

**Comment by brandon-rhodes** [comment, marked as answer]
Alas, no. You will need to create your own function, probably with an `if` statement inside which selects the `EarthSatellite` with the TLE that applies best for the time period you are interested in.

**Reply by davidmikolas** [reply]
This is a really interesting problem that comes up a lot. If we just want the most accurate predictions, then piecewise calculations with transitions that have transitions at the mid-points between consecutive TLE epochs does seem to be the right way to do it, since by definition TLEs are most accurate at their epochs. 

However, If we want to plot an orbit over a period long enough that it spans several TLEs and want it to be at least continuous and smooth, we have to do some dove-tailing, and as soon as we start doing that we run into trouble. I've seen clusters of TLEs so closely spaced its absurd milliseconds and microseconds even, clearly some software glitch or non-optimal behavior. In some cases consecutive TLEs predict positions at their midpoint that differ substantially, and this can be because of a station-keeping maneuver; the satellite actually made a small change in its trajectory. 

So how to manage calculations that span multiple TLEs can be seen to be a user-specific task as different users might have different requirements or use different selection criteria to throw out some bad or erroneous TLEs. 

All that said, a generic dove-tailing (linear weighted average between pairs of points) functionality might be easy enough to write that perhaps it could be mentioned somewhere somehow.

**Reply by syk-hugo** [reply]
Thank you for your answers. It is very interesting, I will try to implement these solutions.

**Reply by CosmicMarkus** [reply]
@brandon-rhodes Firstly, thank you for this amazing software! 

Now, onto the discussion...

Has any progress been made on the idea of a built-in TLE time series object? 

In my opinion, it's a missed opportunity and big barrier to entry for using Skyfield in a professional setting.

The workflow many users want:

1. Get many (tens to hundreds) of TLEs for a satellite over a specified time range 
a. user will provide a text file or interface to a custom database
2. Merge the TLE files into a single EarthSatellite object
3. Let Skyfield work out the logic for piecewise calculations to match the time step with the closest TLE epoch
a. Don't throw out any bad or erroneous TLEs, simply use the data as it is
4. Provide options for matching a TLE with a time period
a. always use the newest TLE available for a time period
b. as @davidmikolas discussed above, calculate values at a mid-point between TLEs
5. I don't want the orbit to be continuous and smooth. The jumps in the data can be very illustrative for showing the limitations of TLEs.

Why do I think this should be built-in to Skyfield as a convenience class? Because if it's not, I'll use other software that makes life easier.

The professional STK package provides this quality-of-life functionality via the `ImportTLEFile` command and `Merge` option (https://help.agi.com/stk/Subsystems/connectCmds/connectCmds.htm#cmd_ImportTLEFileSatellites.htm). 

> Only one (1) satellite is created for each SSC number. Satellites are loaded with all TLEs within the file matching its SSC Number.

Other open-source projects, such as egemenimre/satkit, have a `TleTimeSeries` object (https://satkit.readthedocs.io/en/latest/usagedocs/propagation/tle_storage.html#lists-of-tles-of-the-same-satellite-tletimeseries).

> A specific class [TleTimeSeries](https://satkit.readthedocs.io/en/latest/_autosummary/satkit.propagation.tle_list.html#satkit.propagation.tle_list.TleTimeSeries) exists to store the multiple TLEs from a single satellite with time ordering (epoch values). This is particularly useful to plot the orbit or to feed it to a propagator to propagate the satellite orbit through multiple TLEs.

Sure, with extra effort and time, I might be able to hack something together that will do this in Skyfield. But because I'm not a computer scientist, I'll never be sure I've done it in the most pythonic and efficient manner. And it's an extra hurdle to overcome when I just want to calculate some cool orbital mechanics!
